### PR TITLE
Feature/templatedtracers2

### PR DESCRIPTION
**Pull request by r-englund**
Rebase of #119 onto newest master

### CODE REVIEW

File: include/inviwo/core/util/foreach.h [hunk 1]
```diff
@@ -51,6 +51,19 @@ void foreach_helper(std::true_type, IT a, IT b, Callback callback, size_t start
     std::for_each(a, b, [&](auto v) { callback(v, start++); });
 }
 
+template <typename Callback, typename IT>
+auto foreach_helper_pool(std::true_type, IT a, IT b, Callback callback, size_t start = 0) {
+    return dispatchPool([=]() {
+        auto s = start;
```
Comment (martinfalk) [inline include/inviwo/core/util/foreach.h]: move assignment into capture?

File: modules/vectorfieldvisualization/datastructures/integralline.cpp [hunk 2]
```diff
@@ -115,32 +91,51 @@ double IntegralLine::getLength() const {
...
 }
 
-double IntegralLine::distBetweenPoints(size_t a, size_t b)const {
-    if(a==b) return 0;
-    if(a>b) return distBetweenPoints(b,a);
-    return calcLength( positions_.begin()+a,positions_.begin()+b+1 );
+double IntegralLine::distBetweenPoints(size_t a, size_t b) const {
+    if (a == b) return 0;
+    if (a > b) return distBetweenPoints(b, a);
+    return calcLength(positions_.begin() + a, positions_.begin() + b + 1);
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/datastructures/integralline.cpp:105]: why `begin() + a`  but `begin() + b + 1` ?

File: modules/vectorfieldvisualization/datastructures/integralline.h [hunk 3]
```diff
@@ -44,59 +48,185 @@ namespace inviwo {
...
+const std::vector<T> &IntegralLine::getMetaData(const std::string &name) const {
+    auto it = metaData_.find(name);
+    if (it == metaData_.end()) {
+        throw Exception("No meta data with name: " + name, IvwContext);
+    }
+    auto askedDF = DataFormat<T>::get();
+    auto isDF = it->second->getDataFormat();
+    if (isDF != askedDF) {
+        std::ostringstream oss;
+        oss << "Incorrect dataformat for metadata " << name << " asking for "
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/datastructures/integralline.h]: Metadata is spelled differently in some of the messages, i.e. sometimes with a space in between meta and data, sometimes not.

File: modules/vectorfieldvisualization/datastructures/integralline.h [hunk 4]
```diff
@@ -44,59 +48,185 @@ namespace inviwo {
...
+
+    return static_cast<T>(
+        Interpolation<TV, double>::linear(static_cast<TV>(*prevMD), static_cast<TV>(*nextMD), x));
+}
+
+template <class Elem, class Traits>
+std::basic_ostream<Elem, Traits> &operator<<(std::basic_ostream<Elem, Traits> &os,
+                                             IntegralLine::TerminationReason reason) {
+    switch (reason) {
+        case inviwo::IntegralLine::TerminationReason::StartPoint:
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/datastructures/integralline.h]: remove `inviwo::`

File: modules/vectorfieldvisualization/datastructures/integrallineset.cpp [hunk 5]
```diff
@@ -55,25 +57,30 @@ inviwo::IntegralLine& IntegralLineSet::at(size_t idx) { return lines_.at(idx); }
 
 const inviwo::IntegralLine& IntegralLineSet::at(size_t idx) const { return lines_.at(idx); }
 
-void IntegralLineSet::push_back(IntegralLine &line) {
-    line.setIndex(lines_.size());
-    lines_.push_back(line);
+void IntegralLineSet::push_back(const IntegralLine& line, SetIndex updateIndex) {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/datastructures/integrallineset.cpp:60]: is there a particular reason why `updateIndex` is not a `bool`?

File: modules/vectorfieldvisualization/datastructures/integrallineset.cpp [hunk 6]
```diff
@@ -55,25 +57,30 @@ inviwo::IntegralLine& IntegralLineSet::at(size_t idx) { return lines_.at(idx); }
...
-    lines_.push_back(line);
+void IntegralLineSet::push_back(const IntegralLine& line, size_t idx) {
+    IntegralLine copy(line);
+    copy.setIndex(idx);
+    lines_.push_back(std::move(copy));
 }
 
-void IntegralLineSet::push_back(const IntegralLine &line) {
-    lines_.emplace_back(line);
+void IntegralLineSet::push_back(IntegralLine&& line, SetIndex updateIndex) {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/datastructures/integrallineset.cpp:74]: `bool` for updateIndex?

File: modules/vectorfieldvisualization/integrallinetracer.h [hunk 7]
```diff
@@ -32,34 +32,269 @@
...
+}
+
+template <typename SpatialSampler, typename V, typename H, typename M, typename P>
+P seedTransform(std::false_type, const M &m, const P &pIn) {
+    auto p = m * H(pIn, 1.0f);
+    return P(p) / p[SpatialSampler::DataDimensions];
+}
+
+template <typename P, typename V, typename F>
+P move2(std::true_type, const P &pos, const V &offset, F stepsize) {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/integrallinetracer.h]: `move2`? Do you mean something like "move to"? If so, why not call the function `moveTo`.

File: modules/vectorfieldvisualization/integrallinetracer.h [hunk 8]
```diff
@@ -32,34 +32,269 @@
...
 #include <modules/vectorfieldvisualization/properties/integrallineproperties.h>
+#include <modules/vectorfieldvisualization/datastructures/integralline.h>
+
+#include <unordered_map>
 
 namespace inviwo {
 
+namespace detail {
+
+template <typename SpatialSampler, typename V, typename H, typename M, typename P>
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/integrallinetracer.h:47]: any chance of comments?

File: modules/vectorfieldvisualization/integrallinetracer.h [hunk 9]
```diff
@@ -32,34 +32,269 @@
...
+}  // namespace detail
+
 /**
  * \class IntegralLineTracer
  * \brief VERY_BRIEFLY_DESCRIBE_THE_CLASS
  * DESCRIBE_THE_CLASS
  */
-class IVW_MODULE_VECTORFIELDVISUALIZATION_API IntegralLineTracer { 
+template <typename SpatialSampler,
+          bool TimeDependent = SpatialSampler::SpatialDimensions != SpatialSampler::DataDimensions>
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/integrallinetracer.h:117]: Reasoning behind this check?

File: modules/vectorfieldvisualization/integrallinetracer.h [hunk 10]
```diff
@@ -32,34 +32,269 @@
...
+    };
+
+    const static bool IsTimeDependent = TimeDependent;
+
+    /*
+     * Various types used within this class
+     */
+    using SpatialVector = Vector<SpatialSampler::SpatialDimensions, double>;
+    using DataVector = Vector<SpatialSampler::DataDimensions, double>;
+    using DataHomogenousVector = Vector<SpatialSampler::DataDimensions + 1, double>;
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/integrallinetracer.h:135]: can be a bit tricky when dealing with 2D coords for example.

File: modules/vectorfieldvisualization/integrallinetracer.h [hunk 11]
```diff
@@ -32,34 +32,269 @@
...
+    using SpatialMatrix = Matrix<SpatialSampler::SpatialDimensions, double>;
+    using DataMatrix = Matrix<SpatialSampler::DataDimensions, double>;
+    using DataHomogenouSpatialMatrixrix = Matrix<SpatialSampler::DataDimensions + 1, double>;
+
+    IntegralLineTracer(std::shared_ptr<const Sampler> sampler,
+                       const IntegralLineProperties &properties)
+        : integrationScheme_(properties.getIntegrationScheme())
+        , steps_(properties.getNumberOfSteps())
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/integrallinetracer.h]: implementations should be moved out of the class definition.

File: modules/vectorfieldvisualization/processors/2d/seedpointgenerator2d.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace inviwo {
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming scheme
+const ProcessorInfo SeedPointGenerator2D::processorInfo_{
+    "org.inviwo.SeedPointGenerator2D",  // Class identifier
+    "Seed Point Generator 2D",          // Display name
+    "Data Creation",                    // Category
+    CodeState::Experimental,            // Code state
+    Tags::None,                         // Tags
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/2d/seedpointgenerator2d.cpp]: no tags? How about `"CPU, Seed Points, Generator"`

File: modules/vectorfieldvisualization/processors/3d/pathlines.h [hunk 13]
```diff
@@ -36,36 +36,25 @@
...
- * ### Outports
- *   * __<Outport1>__ <description>.
- *
- * ### Properties
- *   * __<Prop1>__ <description>.
- *   * __<Prop2>__ <description>
- */
-class IVW_MODULE_VECTORFIELDVISUALIZATION_API PathLines : public Processor {
+
+class IVW_MODULE_VECTORFIELDVISUALIZATION_API PathLinesDeprecated : public Processor {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/3d/pathlines.h:53]: kept for compatibility reasons?
maybe consider using `[[deprecated]]` warnings

File: modules/vectorfieldvisualization/processors/3d/pathlines.cpp [hunk 14]
```diff
@@ -34,20 +34,21 @@
...
 namespace inviwo {
 
-const ProcessorInfo PathLines::processorInfo_{
-    "org.inviwo.PathLines",        // Class identifier
-    "Path Lines",                  // Display name
-    "Vector Field Visualization",  // Category
-    CodeState::Experimental,       // Code state
-    Tags::CPU,                     // Tags
+const ProcessorInfo PathLinesDeprecated::processorInfo_{
+    "org.inviwo.PathLinesDepricated",  // Class identifier
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/3d/pathlines.cpp]: typo "deprecated"

File: modules/vectorfieldvisualization/processors/3d/pathlines.cpp [hunk 15]
```diff
@@ -34,20 +34,21 @@
...
+    "Path Lines",                      // Display name
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/3d/pathlines.cpp]: add `" (deprecated)" to display name

File: modules/vectorfieldvisualization/processors/3d/streamlines.cpp [hunk 16]
```diff
@@ -1,42 +1,81 @@
...
 namespace inviwo {
 
-const ProcessorInfo StreamLines::processorInfo_{
-    "org.inviwo.StreamLines",      // Class identifier
-    "Stream Lines",                // Display name
-    "Vector Field Visualization",  // Category
-    CodeState::Experimental,       // Code state
-    Tags::CPU,                     // Tags
+const ProcessorInfo StreamLinesDeprecated::processorInfo_{
+    "org.inviwo.StreamLinesDepricated",  // Class identifier
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/3d/streamlines.cpp]: typo deprecated

File: modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+        tracer.addMetaDataSampler(key, meta.second);
+    }
+
+    std::mutex mutex;
+    size_t startID = 0;
+    for (const auto &seeds : seeds_) {
+        util::forEachParallel(*seeds, [&](const auto &p, size_t i) {
+            IntegralLine line = tracer.traceFrom(p);
+            auto size = line.getPositions().size();
+            if (size > 1) {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h:124]: so the very first line never gets pushed?

File: modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+template <>
+struct ProcessorTraits<StreamLines3D> {
+    static ProcessorInfo getProcessorInfo() {
+        return {
+            "org.inviwo.StreamLines3D",  // Class identifier
+            "Stream Lines 3D",           // Display name
+            "Integral Line Tracer",      // Category
+            CodeState::Experimental,     // Code state
+            Tags::CPU                    // Tags
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h]: maybe more tags?

File: modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+template <>
+struct ProcessorTraits<PathLines3D> {
+    static ProcessorInfo getProcessorInfo() {
+        return {
+            "org.inviwo.PathLines3D",  // Class identifier
+            "Path Lines 3D",           // Display name
+            "Integral Line Tracer",    // Category
+            CodeState::Experimental,   // Code state
+            Tags::CPU                  // Tags
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h]: maybe more tags?

File: modules/vectorfieldvisualization/processors/integrallinevectortomesh.cpp [hunk 20]
```diff
@@ -32,55 +32,101 @@
...
+    typename F = typename std::conditional<std::is_same<T, float>::value, float, double>::type>
+static double norm(const glm::vec<L, T, Q> &glm) {
+    return glm::length(util::glm_convert<glm::vec<L, F, Q>>(glm));
+}
+}  // namespace detail
+
+PropertyClassIdentifier(IntegralLineVectorToMesh::ColorByPropertiy,
+                        "org.inviwo.IntegralLineVectorToMesh.ColorByProperty");
+
+IntegralLineVectorToMesh::ColorByPropertiy::ColorByPropertiy(
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinevectortomesh.cpp]: typo `ColorByPropertiy`

File: modules/vectorfieldvisualization/processors/integrallinevectortomesh.h [hunk 21]
```diff
@@ -48,62 +49,88 @@
...
- */
-class IVW_MODULE_VECTORFIELDVISUALIZATION_API IntegralLineVectorToMesh : public Processor { 
+class IVW_MODULE_VECTORFIELDVISUALIZATION_API IntegralLineVectorToMesh : public Processor {
 public:
-    enum class ColoringMethod {
-        Velocity,
-        Timestamp,
-        ColorPort,
-        Curvature
+    class ColorByPropertiy : public CompositeProperty {
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinevectortomesh.h]: typo `ColorByPropertiy`

File: modules/vectorfieldvisualization/processors/integrallinevectortomesh.h [hunk 22]
```diff
@@ -48,62 +49,88 @@
...
+        DoubleProperty minValue_;
+        DoubleProperty maxValue_;
+        TransferFunctionProperty tf_;
+
+        std::string key_;
     };
+
+    enum class Output { Lines, Ribbons };
+
+    enum class BrushBy { Never, LineIndex, VectorPosition };
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/processors/integrallinevectortomesh.h]: consider using `Nothing` instead of `Never`

File: modules/vectorfieldvisualization/properties/pathlineproperties.cpp [hunk 23]
```diff
@@ -16,8 +16,7 @@
  *
  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
  * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
- * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A P OR CONTRIBUTORS BE LIABLE FOR
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/properties/pathlineproperties.cpp]: what happened here?

File: modules/vectorfieldvisualization/vectorfieldvisualizationmodule.cpp [hunk 24]
```diff
@@ -204,11 +215,54 @@ bool VectorFieldVisualizationModule::Converter::convert(TxElement* root) {
             res |= xml::changeIdentifiers(root, repl);
             return res;
         }
+        case 1: {
+            for (const auto& fromTO : std::vector<std::pair<std::string, std::string>>{
+                     {"StreamLines", "StreamLinesDepricated"},
```
Comment (martinfalk) [inline modules/vectorfieldvisualization/vectorfieldvisualizationmodule.cpp]: typos `...Depricated`

File: src/qt/editor/helpwidget.cpp [hunk 25]
```diff
@@ -264,9 +264,25 @@ QVariant HelpBrowser::loadResource(int type, const QUrl& name) {
             QImage resized{
                 image.scaled(std::max(200, width() - 60), image.height(), Qt::KeepAspectRatio)};
 
+            if (name.toString().contains("form_")) {
```
Comment (martinfalk) [inline src/qt/editor/helpwidget.cpp:267]: is this handling formulas? Maybe add a comment.

**Review by martinfalk — CHANGES_REQUESTED** [review]
nice work. Some minor things left to be done.

File: modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h (hunk 17 above)
Comment (r-englund) [reply modules/vectorfieldvisualization/processors/integrallinetracerprocessor.h:124]: Size refers to number of points in the line, it has to have 2 or more point to be a line :)

File: src/qt/editor/helpwidget.cpp (hunk 25 above)
Comment (r-englund) [reply src/qt/editor/helpwidget.cpp:267]: Maybe I should remove this commit completely due to #181

File: modules/vectorfieldvisualization/properties/pathlineproperties.cpp (hunk 23 above)
Comment (r-englund) [reply modules/vectorfieldvisualization/properties/pathlineproperties.cpp]: I guess I accidentally hit a key-combination to remove a line

File: modules/vectorfieldvisualization/datastructures/integrallineset.cpp (hunk 5 above)
Comment (r-englund) [reply modules/vectorfieldvisualization/datastructures/integrallineset.cpp:60]: Yes, to make the code more readable. 
From the Google C++ Style Guide ([Implementation Comments § Function Argument Comments](https://google.github.io/styleguide/cppguide.html#Implementation_Comments))
 > Consider changing the function signature to replace a bool argument with an enum argument. This will make the argument values self-describing.

File: modules/vectorfieldvisualization/datastructures/integralline.cpp (hunk 2 above)
Comment (r-englund) [reply modules/vectorfieldvisualization/datastructures/integralline.cpp:105]: We want the distance between index a and b, including index b

File: modules/vectorfieldvisualization/datastructures/integrallineset.cpp (hunk 6 above)
Comment (petersteneteg) [reply modules/vectorfieldvisualization/datastructures/integrallineset.cpp:74]: Agree with rickard, in the ideal world one should almost never use bool parameters...

File: include/inviwo/core/util/foreach.h (hunk 1 above)
Comment (r-englund) [reply include/inviwo/core/util/foreach.h]: If a change to this I get `error: increment of read-only location ‘id’`
```c++
return dispatchPool([id=start,callback,a,b]() {
    std::for_each(a, b, [&](auto v) { callback(v, id++); });
});
```
(https://stackoverflow.com/a/27715105/600633)

Comment (petersteneteg) [reply include/inviwo/core/util/foreach.h]: there is nothing wrong about making the lambda mutable, (all member of the lambda is const by default, by adding mutable they are no longer const) what is the problem? 
Think mating wanted to move the callback into the lambda to avoid the copy...
```
return dispatchPool([id=start, c = std::move(callback), a, b]() mutable {
    std::for_each(a, b, [&](auto v) { c(v, id++); });
});
```

File: modules/vectorfieldvisualization/datastructures/integrallineset.cpp (hunk 5 above)
Comment (martinfalk) [reply modules/vectorfieldvisualization/datastructures/integrallineset.cpp:60]: good suggestion actually

File: include/inviwo/core/util/foreach.h (hunk 1 above)
Comment (martinfalk) [reply include/inviwo/core/util/foreach.h]: I was actually thinking of only capturing `s = start`

**Review by martinfalk — APPROVED** [review]

File: include/inviwo/core/processors/processorstate.h [hunk 26]
```diff
@@ -35,7 +35,7 @@
 
 namespace inviwo {
 
-enum class CodeState { Broken, Experimental, Stable };
+enum class CodeState { Broken, Experimental, Stable, Deprecated };
```
Comment (petersteneteg) [inline include/inviwo/core/processors/processorstate.h:38]: This should be in a different PR, have you also added icons and updated the GUI?

Comment (r-englund) [reply include/inviwo/core/processors/processorstate.h:38]: Yes, the gui is update,  @martinfalk helped me with the icon. 
image

Comment (petersteneteg) [reply include/inviwo/core/processors/processorstate.h:38]: 👍
